source/common/router/router.cc

```diff
@@ -345,7 +345,7 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   ASSERT(headers.Method());
   ASSERT(headers.Host());
 
-  downstream_headers_ = &headers;
+  downstream_headers_ = std::make_unique<Http::HeaderMapImpl>(headers);
```

Comment:
Good catch on use-after deletion!  I'm a bit leery of the copy from a perf perspective.  If Matt thinks this is the right thing to do from an expediency standpoint how about a TODO to see if we can sort out the underlying lifetime issues later?

Reply:
We should be logging before teardown. What is the failing call stack?

Reply:
[2019-10-31 17:53:38.259][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:83] Caught Segmentation fault, suspect faulting address 0x2337de
[2019-10-31 17:53:38.259][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[2019-10-31 17:53:38.259][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 0/1.12.0-dev/redacted/RELEASE/BoringSSL
[2019-10-31 17:53:38.259][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: __restore_rt [0x7f9428177390]
[2019-10-31 17:53:38.261][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #1: Envoy::Router::Filter::UpstreamRequest::~UpstreamRequest() [0x7cefde]
[2019-10-31 17:53:38.263][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: Envoy::Router::Filter::resetAll() [0x7caa9f]
[2019-10-31 17:53:38.264][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #3: Envoy::Router::Filter::onDestroy() [0x7ca9ae]
[2019-10-31 17:53:38.266][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #4: Envoy::Http::AsyncStreamImpl::reset() [0x7c1702]
[2019-10-31 17:53:38.267][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #5: Envoy::Http::AsyncClientImpl::~AsyncClientImpl() [0x7bfb22]
[2019-10-31 17:53:38.269][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #6: Envoy::Http::(anonymous namespace)::AsyncClientImplTest::~AsyncClientImplTest() [0x525b89]
[2019-10-31 17:53:38.270][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #7: Envoy::Http::(anonymous namespace)::AsyncClientImplTest_WatermarkCallbacks_Test::~AsyncClientImplTest_WatermarkCallbacks_Test() [0x570ace]
[2019-10-31 17:53:38.272][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #8: testing::internal::HandleExceptionsInMethodIfSupported<>() [0xc313d4]
[2019-10-31 17:53:38.274][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #9: testing::TestInfo::Run() [0xc32144]
[2019-10-31 17:53:38.275][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #10: testing::TestSuite::Run() [0xc32a97]

Reply:
I'm honestly inclined to use the manual byte size count in the logger for this headermap and add as a TODO to remove ASAP so at least it doesn't crash in the release.

Reply:
From a very quick look this is an order of destuction issue inside async client, and I think the bug is there. I would rather not add the copy to the router.

If there is a quick/slow/localized fix only for the upstream logging, let's just do that w/ a TODO?

Reply:
It won't be any slower to call the manual byte size call inside the logger, since the refresh outside the logger does exactly that too.

I'll investigate what the destruction issue is.

Reply:
OK yeah we can fix async client later. The issue basically is that in the destroy/reset case we are destroying the headers before the stream in the AsyncRequest case, and possibly not even storing them for the streaming case.